spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/core/subscriber/PubSubSubscriberOperations.java

```diff
@@ -84,27 +85,32 @@
 	PubsubMessage pullNext(String subscription);
 
 	/**
-	 * Acknowledge a batch of messages. The method will group the messages by subscription name
-	 * and acknowledge them in batches.
+	 * Acknowledge a batch of messages. The messages must have the same project id and be from the same subscription.
```

Comment:
> The messages must have the same project id and be from the same subscription

Why? What is the point then have an argument as as `Collection<AcknowledgeablePubsubMessage>`?
Just add two more args: `projectId` and `subscription`. And request just `Collection<String> ackIds`.

What am I missing? What was wrong with grouping?

Thanks

Reply:
Given the follow statements ...

1. The `PubSubSubscriberTemplate` requires a `SubscriberFactory`.
2. The `SubscriberFactory` has a `projectId` and requires a `CredentialsProvider`.
3. The GCP Pub / Sub READ role for a subscription can be specific to a user / service account, and it not project global.

This would require any `Collection<AcknowledgeablePubsubMessage>` to contain only messages which have the same `projectId` and subscription name so that they match the `PubSubSubscriberTemplate` (or likewise configured `PubSubSubscriptionTemplate`) that generated the messages. 

The method *could* be changed to accept a `Collection<String> ackIds`, as well as as `projectId` and `subscription` ... but given the `PubSubSubscriberTemplate` deals with `AcknowledgeablePubsubMessage`, in my opinion it is much cleaner and developer friendly to accept `Collection<AcknowledgeablePubsubMessage>` and have the method deal with the work to collection the `Collection<String> ackIds`.